### PR TITLE
Re-use cache key from execution history

### PR DESCRIPTION
**Pull request by wolfs**
To avoid re-calculating the cache key when the task is up-to-date.

Instead of calculating the cache key in `CaptureStateBeforeExecutionStep`, we first do up-to-date checks. If the task is up-to-date, we use the cache key from the previous execution, and if it isn’t, we re-calculate the key. This allows us to then store the cache key in the execution history after the task executed.
The reason for doing it like this is that up-to-date checks are fast and calculating the cache key is slow.
Note that for the non-incremental pipeline we don’t try to re-use the cache key from the execution history - since we don’t have any - and calculate it as before in a `NonIncrementalResolveCachingStateStep`.

### CODE REVIEW

File: platforms/core-execution/execution/src/main/java/org/gradle/internal/execution/steps/IncrementalChangesContext.java [hunk 1]
```diff
@@ -47,4 +59,9 @@ public ImmutableList<String> getRebuildReasons() {
     public Optional<ExecutionStateChanges> getChanges() {
         return Optional.ofNullable(executionStateChanges);
     }
+
+    @Nullable
+    public HashCode getCacheKey() {
+        return cacheKey;
+    }
```
Comment (lptr) [inline platforms/core-execution/execution/src/main/java/org/gradle/internal/execution/steps/IncrementalChangesContext.java:66]: :x: This should return `Optional<HashCode>`.

File: platforms/core-execution/execution/src/main/java/org/gradle/internal/execution/steps/IncrementalResolveCachingStateStep.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+import org.gradle.caching.internal.controller.BuildCacheController;
+import org.gradle.internal.execution.UnitOfWork;
+import org.gradle.internal.execution.caching.CachingState;
+import org.gradle.internal.hash.HashCode;
+
+import java.util.Objects;
+import java.util.Optional;
+
+public class IncrementalResolveCachingStateStep<C extends IncrementalChangesContext> extends AbstractResolveCachingStateStep<C> {
```
Comment (lptr) [inline platforms/core-execution/execution/src/main/java/org/gradle/internal/execution/steps/IncrementalResolveCachingStateStep.java:27]: 🤔 While there would be benefits to having `Incremental` / `NonIncremental` be a prefix of the step name always, we currently don't do that with the others, and instead aim for a more readable name. I think we should go with `ResolveIncrementalCachingStateStep` here as well.

File: platforms/core-execution/execution/src/main/java/org/gradle/internal/execution/steps/ResolveChangesStep.java [hunk 3]
```diff
@@ -22,51 +22,75 @@
...
 
         return delegate.execute(work, delegateContext);
     }
 
+    private HashCode calculateCacheKey(ExecutionStateChanges changes) {
+        Logger logger = emitBuildCacheDebugLogging.getAsBoolean()
+            ? LOGGER
+            : NOPLogger.NOP_LOGGER;
+        return new DefaultCachingStateFactory(logger).calculateCacheKey(changes.getBeforeExecutionState());
+    }
```
Comment (lptr) [inline platforms/core-execution/execution/src/main/java/org/gradle/internal/execution/steps/ResolveChangesStep.java:90]: :x: This should only receive the `BeforeExecutionState`.

File: platforms/core-execution/execution/src/main/java/org/gradle/internal/execution/steps/StoreExecutionStateStep.java [hunk 4]
```diff
@@ -47,7 +49,7 @@ public R execute(UnitOfWork work, C context) {
                 .flatMap(beforeExecutionState -> result.getAfterExecutionOutputState()
                     .filter(afterExecutionState -> result.getExecution().isSuccessful() || shouldPreserveFailedState(context, afterExecutionState))
                     .map(executionOutputState -> new DefaultAfterExecutionState(beforeExecutionState, executionOutputState)))
-                .ifPresent(afterExecutionState -> history.store(context.getIdentity().getUniqueId(), afterExecutionState)));
+                .ifPresent(afterExecutionState -> history.store(context.getIdentity().getUniqueId(), Objects.requireNonNull(context.getCacheKey()), afterExecutionState)));
```
Comment (lptr) [inline platforms/core-execution/execution/src/main/java/org/gradle/internal/execution/steps/StoreExecutionStateStep.java:52]: 🤔 This is the part I'm unhappy about, but let's encode the "no cache key available" case in the context type hierarchy at a later time. Could you please add a TODO here?

File: platforms/core-execution/execution/src/test/groovy/org/gradle/internal/execution/steps/BuildCacheStepTest.groovy [hunk 5]
```diff
@@ -308,3 +308,9 @@ class BuildCacheStepTest extends StepSpec<IncrementalChangesContext> implements
         1 * buildCacheController.store(cacheKey, _, outputFilesProducedByWork, Duration.ofMillis(123L)) >> { storeResult() }
     }
 }
+
+abstract class MyCachingContext extends WorkspaceContext implements CachingContext {
+    MyCachingContext(WorkspaceContext parent) {
+        super(parent)
+    }
+}
```
Comment (lptr) [inline platforms/core-execution/execution/src/test/groovy/org/gradle/internal/execution/steps/BuildCacheStepTest.groovy:316]: ❓ Why is this needed?

**Review by lptr — COMMENTED** [review]
LGTM except for a few small things.

File: platforms/core-execution/execution/src/main/java/org/gradle/internal/execution/steps/IncrementalChangesContext.java (hunk 1 above)
Comment (wolfs) [reply platforms/core-execution/execution/src/main/java/org/gradle/internal/execution/steps/IncrementalChangesContext.java:69]: Done.

File: platforms/core-execution/execution/src/main/java/org/gradle/internal/execution/steps/IncrementalResolveCachingStateStep.java (hunk 2 above)
Comment (wolfs) [reply platforms/core-execution/execution/src/main/java/org/gradle/internal/execution/steps/IncrementalResolveCachingStateStep.java]: Renamed.

File: platforms/core-execution/execution/src/main/java/org/gradle/internal/execution/steps/ResolveChangesStep.java (hunk 3 above)
Comment (wolfs) [reply platforms/core-execution/execution/src/main/java/org/gradle/internal/execution/steps/ResolveChangesStep.java:90]: Done.

File: platforms/core-execution/execution/src/main/java/org/gradle/internal/execution/steps/StoreExecutionStateStep.java (hunk 4 above)
Comment (wolfs) [reply platforms/core-execution/execution/src/main/java/org/gradle/internal/execution/steps/StoreExecutionStateStep.java]: I added a TODO.

**Review by ov7a — APPROVED** [review]

**Review by lptr — APPROVED** [review]
LGTM.

**Comment by wolfs** [comment]
@bot-gradle merge
